### PR TITLE
resolves #1311: constrain the amount of enumerations the new planner …

### LINKED ISSUE #1311: constrain the amount of enumerations the new planner can do
Currently, enumerations of matches per rule call are not constraint in any way. A query that causes many variations to be created for a particular group could cause performance issues due to over enumeration. While hypothetical cases like that should be fixed for each individual case, this issue should just address a fail safe to abort planning when over enumeration occurs.

### PR DESCRIPTION
**Pull request by normen662**
…can do

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java [hunk 1]
```diff
@@ -348,6 +349,16 @@ public void setMaxTotalTaskCount(final int maxTotalTaskCount) {
                 .build();
     }
 
+    /**
+     * Set the maximum number of yields that are permitted per rule call within the Cascades planner.
+     * @param maxNumYieldsPerRuleCall the desired maximum number of yields that are permitted per rule call
```
Comment (ohadzeliger) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java:354]: Add statement about '0' being "unrestricted"

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java [hunk 2]
```diff
@@ -668,18 +683,21 @@ protected boolean shouldExecute() {
...
             final PlannerBindings initialBindings = getInitialBindings();
 
+            final AtomicInteger numMatches = new AtomicInteger(0);
+
             rule.getMatcher()
                     .bindMatches(initialBindings, getBindable())
                     .map(bindings -> new CascadesRuleCall(getContext(), rule, group, aliasResolver, bindings))
                     .forEach(ruleCall -> {
+                        if (isMaxNumMatchesPerRuleCallExceeded(configuration, numMatches.incrementAndGet())) {
+                            throw new RecordQueryPlanComplexityException("Maximum number of matches per rule call (" + configuration.getMaxNumMatchesPerRuleCall() + ") is exceeded");
```
Comment (ohadzeliger) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java:699]: Add the offending rule?

**Review by ohadzeliger — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java (hunk 1 above)
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java:354]: done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java (hunk 2 above)
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java:699]: done

**Review by ohadzeliger — APPROVED** [review]

**Pull request merged**